In [1]:
import numpy as np
import pandas as pd

## Koşullu Seçimler

In [2]:
df = pd.DataFrame({'id':[1001, 1002, 1003, 1004, 1005, 1006, 1007, 1008, 1009],
                   'name':['ali', 'ayse', 'fatma', 'murat', 'hasan', 'betul', 'berke', 'nuran', 'nese'],
                   'surname':['guzel', 'can', 'yilmaz', 'kaya', 'demir', 'sahin', 'celik', 'aydin', 'aslan'],
                   'gender':['E', 'K', 'K', 'E', 'E', 'K', 'E', 'K', 'K'],
                   'city':['istanbul', 'izmir', 'ankara', 'istanbul', 'izmir', 'izmir', 'antalya', 'edirne', 'ankara'],
                   'year':[1990, 1995, 1992, 1980, 2000, 2005, 2010, 1999, 1985],
                   'price':[5000, 3000, 2000, 7000, 14000, 6000, 2600, 7500, 8300]})
df

id   name surname gender      city  year  price
0  1001    ali   guzel      E  istanbul  1990   5000
1  1002   ayse     can      K     izmir  1995   3000
2  1003  fatma  yilmaz      K    ankara  1992   2000
3  1004  murat    kaya      E  istanbul  1980   7000
4  1005  hasan   demir      E     izmir  2000  14000
5  1006  betul   sahin      K     izmir  2005   6000
6  1007  berke   celik      E   antalya  2010   2600
7  1008  nuran   aydin      K    edirne  1999   7500
8  1009   nese   aslan      K    ankara  1985   8300

In [3]:
kosul = df['city'] == 'izmir'
df[kosul]

id   name surname gender   city  year  price
1  1002   ayse     can      K  izmir  1995   3000
4  1005  hasan   demir      E  izmir  2000  14000
5  1006  betul   sahin      K  izmir  2005   6000

In [4]:
df['city'] == 'izmir'  # Koşul sağlanıyorsa True, sağlanmıyorsa False döndürür.

0    False
1     True
2    False
3    False
4     True
5     True
6    False
7    False
8    False
Name: city, dtype: bool

and operatörü için & 
or operatörü için | kullanıyoruz.

In [5]:
df[(df['city'] == 'izmir') | (df['city'] == 'istanbul')]

id   name surname gender      city  year  price
0  1001    ali   guzel      E  istanbul  1990   5000
1  1002   ayse     can      K     izmir  1995   3000
3  1004  murat    kaya      E  istanbul  1980   7000
4  1005  hasan   demir      E     izmir  2000  14000
5  1006  betul   sahin      K     izmir  2005   6000

in operatörünü şu şekilde kullanıyoruz;

In [6]:
df['city'].isin(['izmir', 'istanbul'])

0     True
1     True
2    False
3     True
4     True
5     True
6    False
7    False
8    False
Name: city, dtype: bool

In [7]:
df[(df['city'] == 'izmir') & (df['price']>5000)]

id   name surname gender   city  year  price
4  1005  hasan   demir      E  izmir  2000  14000
5  1006  betul   sahin      K  izmir  2005   6000

In [8]:
~(df['city'] == 'izmir')  # ~ işareti değerin tersini döndürür.

0     True
1    False
2     True
3     True
4    False
5    False
6     True
7     True
8     True
Name: city, dtype: bool

In [3]:
df[~(df['city'] == 'izmir') & (df['price']>5000)]

id   name surname gender      city  year  price
3  1004  murat    kaya      E  istanbul  1980   7000
7  1008  nuran   aydin      K    edirne  1999   7500
8  1009   nese   aslan      K    ankara  1985   8300

In [4]:
df.query("year>2000 or city == 'istanbul'")

id   name surname gender      city  year  price
0  1001    ali   guzel      E  istanbul  1990   5000
3  1004  murat    kaya      E  istanbul  1980   7000
5  1006  betul   sahin      K     izmir  2005   6000
6  1007  berke   celik      E   antalya  2010   2600

In [5]:
df.loc[(df['city'] == 'izmir') & (df['price']>5000), ['name', 'surname']]

name surname
4  hasan   demir
5  betul   sahin

## Dropna - Eksik Verileri Kaldırmak İçin Kullanılır

In [6]:
np.random.seed(35)
df = pd.DataFrame({
    'Date' : pd.date_range(start='2023-01-13',periods=10,freq='D'),
    'Urun' : 222,
    'Deger_1': np.random.randint(1,10,10),
    'Deger_2': np.random.randint(1,10,10),
    'Deger_3': np.random.randint(1,10,10),
    'Deger_4': np.random.randint(1,10,10)
    
})
df

Date  Urun  Deger_1  Deger_2  Deger_3  Deger_4
0 2023-01-13   222        8        3        1        2
1 2023-01-14   222        2        6        2        4
2 2023-01-15   222        1        4        1        7
3 2023-01-16   222        9        4        3        5
4 2023-01-17   222        9        4        3        3
5 2023-01-18   222        9        9        6        4
6 2023-01-19   222        8        8        9        2
7 2023-01-20   222        8        6        4        4
8 2023-01-21   222        9        9        1        5
9 2023-01-22   222        1        7        5        3

In [7]:
df.loc[[2,6,9],"Urun"] = np.nan
df.loc[[4,6],"Deger_1"] = np.nan
df.loc[[4,6],"Deger_2"] = np.nan
df.loc[6:,"Deger_3"] = np.nan
df.loc[3:6,"Deger_4"] = np.nan
df

Date   Urun  Deger_1  Deger_2  Deger_3  Deger_4
0 2023-01-13  222.0      8.0      3.0      1.0      2.0
1 2023-01-14  222.0      2.0      6.0      2.0      4.0
2 2023-01-15    NaN      1.0      4.0      1.0      7.0
3 2023-01-16  222.0      9.0      4.0      3.0      NaN
4 2023-01-17  222.0      NaN      NaN      3.0      NaN
5 2023-01-18  222.0      9.0      9.0      6.0      NaN
6 2023-01-19    NaN      NaN      NaN      NaN      NaN
7 2023-01-20  222.0      8.0      6.0      NaN      4.0
8 2023-01-21  222.0      9.0      9.0      NaN      5.0
9 2023-01-22    NaN      1.0      7.0      NaN      3.0

In [8]:
df.isnull() # Null değerleri gösterir. Ya da df.isna()

Date   Urun  Deger_1  Deger_2  Deger_3  Deger_4
0  False  False    False    False    False    False
1  False  False    False    False    False    False
2  False   True    False    False    False    False
3  False  False    False    False    False     True
4  False  False     True     True    False     True
5  False  False    False    False    False     True
6  False   True     True     True     True     True
7  False  False    False    False     True    False
8  False  False    False    False     True    False
9  False   True    False    False     True    False

In [9]:
df.isnull().sum() 

Date       0
Urun       3
Deger_1    2
Deger_2    2
Deger_3    4
Deger_4    4
dtype: int64

** Satırlardaki toplam null değer sayısını döndürür. Toplam null değer sayısını ise kod sonuna .sum() ekleyerek bulabiliriz.

In [10]:
df.count() # Kaç tane null olmayan değer olduğunu döndürür.

Date       10
Urun        7
Deger_1     8
Deger_2     8
Deger_3     6
Deger_4     6
dtype: int64

In [11]:
df.dropna()

Date   Urun  Deger_1  Deger_2  Deger_3  Deger_4
0 2023-01-13  222.0      8.0      3.0      1.0      2.0
1 2023-01-14  222.0      2.0      6.0      2.0      4.0

** İçerisinde en az 1 tane null olan değer olan satırları kaldırır.

df.dropna()
axis=0 : Satırları 
axis=1 : Sütunları kapsar.

In [12]:
df.dropna(axis=1)

Date
0 2023-01-13
1 2023-01-14
2 2023-01-15
3 2023-01-16
4 2023-01-17
5 2023-01-18
6 2023-01-19
7 2023-01-20
8 2023-01-21
9 2023-01-22

### Bütün değerleri null olan satırı kaldırmak için;

In [13]:
df.dropna(how= 'all')

Date   Urun  Deger_1  Deger_2  Deger_3  Deger_4
0 2023-01-13  222.0      8.0      3.0      1.0      2.0
1 2023-01-14  222.0      2.0      6.0      2.0      4.0
2 2023-01-15    NaN      1.0      4.0      1.0      7.0
3 2023-01-16  222.0      9.0      4.0      3.0      NaN
4 2023-01-17  222.0      NaN      NaN      3.0      NaN
5 2023-01-18  222.0      9.0      9.0      6.0      NaN
6 2023-01-19    NaN      NaN      NaN      NaN      NaN
7 2023-01-20  222.0      8.0      6.0      NaN      4.0
8 2023-01-21  222.0      9.0      9.0      NaN      5.0
9 2023-01-22    NaN      1.0      7.0      NaN      3.0

## Fillna - Eksik Değerleri Doldurma

In [3]:
data = {'isim':['Zeynep','Yusuf','Elif','Ömer','Hamza','Asya','Fatma','Murat','Ahmet','Defne','Can','Berk',np.nan],
       'sirket':['Tesla','Apple','Amazon','Amazon','Tesla','Tesla','Apple','Apple',np.nan,'Amazon','Tesla','Amazon','Apple'],
       'departman':['IT','Sales',np.nan,'Marketing','IT','IT','IT',np.nan,'Marketing','Marketing',np.nan,'Sales','Sales'],
       'maas':[100,150,200,np.nan,np.nan,np.nan,230,np.nan,225,115,160,250,220],
       'prim':[np.nan,np.nan, 10,20,35,34,22,25,40,10,np.nan,35,20]}
df = pd.DataFrame(data)
df

isim  sirket  departman   maas  prim
0   Zeynep   Tesla         IT  100.0   NaN
1    Yusuf   Apple      Sales  150.0   NaN
2     Elif  Amazon        NaN  200.0  10.0
3     Ömer  Amazon  Marketing    NaN  20.0
4    Hamza   Tesla         IT    NaN  35.0
5     Asya   Tesla         IT    NaN  34.0
6    Fatma   Apple         IT  230.0  22.0
7    Murat   Apple        NaN    NaN  25.0
8    Ahmet     NaN  Marketing  225.0  40.0
9    Defne  Amazon  Marketing  115.0  10.0
10     Can   Tesla        NaN  160.0   NaN
11    Berk  Amazon      Sales  250.0  35.0
12     NaN   Apple      Sales  220.0  20.0

#### Aynı değerle doldurma;

In [4]:
df.fillna(0)

isim  sirket  departman   maas  prim
0   Zeynep   Tesla         IT  100.0   0.0
1    Yusuf   Apple      Sales  150.0   0.0
2     Elif  Amazon          0  200.0  10.0
3     Ömer  Amazon  Marketing    0.0  20.0
4    Hamza   Tesla         IT    0.0  35.0
5     Asya   Tesla         IT    0.0  34.0
6    Fatma   Apple         IT  230.0  22.0
7    Murat   Apple          0    0.0  25.0
8    Ahmet       0  Marketing  225.0  40.0
9    Defne  Amazon  Marketing  115.0  10.0
10     Can   Tesla          0  160.0   0.0
11    Berk  Amazon      Sales  250.0  35.0
12       0   Apple      Sales  220.0  20.0

#### Her sütundaki eksik değeri farklı değerle doldurma;

In [7]:
df.fillna({'isim': 'İsim Belirtilmemiş', 'maas':0, 'departman':'Bilinmiyor', 'prim':0, 'sirket':'Bilinmiyor'})

isim      sirket   departman   maas  prim
0               Zeynep       Tesla          IT  100.0   0.0
1                Yusuf       Apple       Sales  150.0   0.0
2                 Elif      Amazon  Bilinmiyor  200.0  10.0
3                 Ömer      Amazon   Marketing    0.0  20.0
4                Hamza       Tesla          IT    0.0  35.0
5                 Asya       Tesla          IT    0.0  34.0
6                Fatma       Apple          IT  230.0  22.0
7                Murat       Apple  Bilinmiyor    0.0  25.0
8                Ahmet  Bilinmiyor   Marketing  225.0  40.0
9                Defne      Amazon   Marketing  115.0  10.0
10                 Can       Tesla  Bilinmiyor  160.0   0.0
11                Berk      Amazon       Sales  250.0  35.0
12  İsim Belirtilmemiş       Apple       Sales  220.0  20.0

In [8]:
new_df = df.copy()
new_df

isim  sirket  departman   maas  prim
0   Zeynep   Tesla         IT  100.0   NaN
1    Yusuf   Apple      Sales  150.0   NaN
2     Elif  Amazon        NaN  200.0  10.0
3     Ömer  Amazon  Marketing    NaN  20.0
4    Hamza   Tesla         IT    NaN  35.0
5     Asya   Tesla         IT    NaN  34.0
6    Fatma   Apple         IT  230.0  22.0
7    Murat   Apple        NaN    NaN  25.0
8    Ahmet     NaN  Marketing  225.0  40.0
9    Defne  Amazon  Marketing  115.0  10.0
10     Can   Tesla        NaN  160.0   NaN
11    Berk  Amazon      Sales  250.0  35.0
12     NaN   Apple      Sales  220.0  20.0

In [9]:
new_df.ffill() # Kendinden önceki değerle doldurur eksik veriyi.

isim  sirket  departman   maas  prim
0   Zeynep   Tesla         IT  100.0   NaN
1    Yusuf   Apple      Sales  150.0   NaN
2     Elif  Amazon      Sales  200.0  10.0
3     Ömer  Amazon  Marketing  200.0  20.0
4    Hamza   Tesla         IT  200.0  35.0
5     Asya   Tesla         IT  200.0  34.0
6    Fatma   Apple         IT  230.0  22.0
7    Murat   Apple         IT  230.0  25.0
8    Ahmet   Apple  Marketing  225.0  40.0
9    Defne  Amazon  Marketing  115.0  10.0
10     Can   Tesla  Marketing  160.0  10.0
11    Berk  Amazon      Sales  250.0  35.0
12    Berk   Apple      Sales  220.0  20.0

In [10]:
new_df.bfill() # Kendinden sonraki değerle doldurur eksik veriyi.

isim  sirket  departman   maas  prim
0   Zeynep   Tesla         IT  100.0  10.0
1    Yusuf   Apple      Sales  150.0  10.0
2     Elif  Amazon  Marketing  200.0  10.0
3     Ömer  Amazon  Marketing  230.0  20.0
4    Hamza   Tesla         IT  230.0  35.0
5     Asya   Tesla         IT  230.0  34.0
6    Fatma   Apple         IT  230.0  22.0
7    Murat   Apple  Marketing  225.0  25.0
8    Ahmet  Amazon  Marketing  225.0  40.0
9    Defne  Amazon  Marketing  115.0  10.0
10     Can   Tesla      Sales  160.0  35.0
11    Berk  Amazon      Sales  250.0  35.0
12     NaN   Apple      Sales  220.0  20.0

In [2]:
import seaborn as sns

In [4]:
df = sns.load_dataset('titanic')

In [5]:
df.rename(columns= {'sex' :'gender'}, inplace=True)

In [6]:
df

survived  pclass  gender   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [7]:
df.head()

survived  pclass  gender   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [8]:
df.tail()

survived  pclass  gender   age  sibsp  parch   fare embarked   class  \
886         0       2    male  27.0      0      0  13.00        S  Second   
887         1       1  female  19.0      0      0  30.00        S   First   
888         0       3  female   NaN      1      2  23.45        S   Third   
889         1       1    male  26.0      0      0  30.00        C   First   
890         0       3    male  32.0      0      0   7.75        Q   Third   

       who  adult_male deck  embark_town alive  alone  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True

In [9]:
df.shape  # [0] : sadece satır bilgisi verir, [1] : sadece sütun bilgisi verir.

(891, 15)

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   gender       891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


In [11]:
df.describe().T

count       mean        std   min      25%      50%   75%       max
survived  891.0   0.383838   0.486592  0.00   0.0000   0.0000   1.0    1.0000
pclass    891.0   2.308642   0.836071  1.00   2.0000   3.0000   3.0    3.0000
age       714.0  29.699118  14.526497  0.42  20.1250  28.0000  38.0   80.0000
sibsp     891.0   0.523008   1.102743  0.00   0.0000   0.0000   1.0    8.0000
parch     891.0   0.381594   0.806057  0.00   0.0000   0.0000   0.0    6.0000
fare      891.0  32.204208  49.693429  0.00   7.9104  14.4542  31.0  512.3292

In [12]:
df[df['age'] > 50]  # Yaşı elliden büyük olanları getirir.

survived  pclass  gender   age  sibsp  parch     fare embarked   class  \
6           0       1    male  54.0      0      0  51.8625        S   First   
11          1       1  female  58.0      0      0  26.5500        S   First   
15          1       2  female  55.0      0      0  16.0000        S  Second   
33          0       2    male  66.0      0      0  10.5000        S  Second   
54          0       1    male  65.0      0      1  61.9792        C   First   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
820         1       1  female  52.0      1      1  93.5000        S   First   
829         1       1  female  62.0      0      0  80.0000      NaN   First   
851         0       3    male  74.0      0      0   7.7750        S   Third   
857         1       1    male  51.0      0      0  26.5500        S   First   
879         1       1  female  56.0      0      1  83.1583        C   First   

       who  adult_male deck  embark_town alive  alone  
6      man        True    E  Southampton    no   True  
11   woman       False    C  Southampton   yes   True  
15   woman       False  NaN  Southampton   yes   True  
33     man        True  NaN  Southampton    no   True  
54     man        True    B    Cherbourg    no  False  
..     ...         ...  ...          ...   ...    ...  
820  woman       False    B  Southampton   yes  False  
829  woman       False    B          NaN   yes   True  
851    man        True  NaN  Southampton    no   True  
857    man        True    E  Southampton   yes   True  
879  woman       False    C    Cherbourg   yes  False  

[64 rows x 15 columns]

In [19]:
df.loc[(df['age']> 50) & (df['gender'] == 'male'), ['survived', 'age']]

survived   age
6           0  54.0
33          0  66.0
54          0  65.0
94          0  59.0
96          0  71.0
116         0  70.5
124         0  54.0
150         0  51.0
152         0  55.5
155         0  51.0
170         0  61.0
174         0  56.0
222         0  51.0
232         0  59.0
249         0  54.0
252         0  62.0
262         0  52.0
280         0  65.0
317         0  54.0
326         0  61.0
406         0  51.0
438         0  64.0
449         1  52.0
456         0  65.0
467         0  56.0
487         0  58.0
492         0  55.0
493         0  71.0
545         0  64.0
555         0  62.0
570         1  62.0
582         0  54.0
587         1  60.0
625         0  61.0
626         0  57.0
630         1  80.0
631         0  51.0
647         1  56.0
659         0  58.0
672         0  70.0
684         0  60.0
694         0  60.0
695         0  52.0
714         0  52.0
745         0  70.0
851         0  74.0
857         1  51.0

In [20]:
df.gender.value_counts()

gender
male      577
female    314
Name: count, dtype: int64

In [23]:
df.groupby('gender').agg({'age':['mean', 'count'],
                          'survived': 'mean'})

age        survived
             mean count      mean
gender                           
female  27.915709   261  0.742038
male    30.726645   453  0.188908

Yani kadınların yaş ortalaması %27, toplam yaş değeri ise 261'dir. Kadınların hayatta kalma ortalaması ise %74'tür. 
Erkeklerin yaş ortalaması %30, toplam yaş değeri ise 453'tür. Erkeklerin hayatta kalma ortalaması ise %18'dir.
İlk bakışta kadınlar daha fazla hayatta kalmıştır yorumu aklımıza gelse de erkek yaş değeri kadın yaş değerinden 192 fazladır. Bunu da değerlendirmeye katmak gerekir. 

In [25]:
df.groupby(['gender', 'class']).agg({'age':['mean', 'count'],
                                     'survived':['mean', 'count']})

age        survived      
                    mean count      mean count
gender class                                  
female First   34.611765    85  0.968085    94
       Second  28.722973    74  0.921053    76
       Third   21.750000   102  0.500000   144
male   First   41.281386   101  0.368852   122
       Second  30.740707    99  0.157407   108
       Third   26.507589   253  0.135447   347

In [26]:
df.pivot_table(index='gender', columns='class', values='age',aggfunc='count')

class   First  Second  Third
gender                      
female     85      74    102
male      101      99    253